### PR TITLE
Support for UnsignedInt & UnsignedBigInt

### PR DESCRIPTION
**Pull request by alexsnaps**
This is nowhere close to be merge'able, but I thought I'd use it to get the discussion going:

While trying to see how Noria would workout on a project, I hit an issue where `unsigned BigInt(20)` column types weren't properly supported. `nom-sql` had issues parsing the SQL while trying to load it all in Noria and threw `ParseIntError { kind: Overflow }` at us. 

I started hacking trying to get something working to at least get to load the the mysql dump into noria and make progress with the testing. Now here I am… `msql-srv` seems to deal with it fine (based on my limited reading of the code), but `nom-sql` less so. I'm more than willing to make progress on this, but would love to know what y'all take is on  adding the `DataType::Unsigned(Big)Int` is. Or would it be nicer to treat these as properties of the DataType itself, just like the "width"? `msql-srv` looks at it as flags on the column definition… Also what about the other "integer" types MySQL supports?

Anyways, I thought I'd get the discussion going already. In the meantime I'll try to make progress here… somehow. Tomorrow tho probably :)

wdyt? Or did I miss something entirely?

### CODE REVIEW

**Comment by ms705** [comment]
This is an entirely reasonable question! :+1:

We can help with `nom-sql` parse failures if you let us know what queries it barfs at. We certainly would love it if there was a utility to read MySQL dumps into Noria!

One perhaps important thing to note is that `DataType` exists in order to allow Noria's operators to be generic over the data they consume and store in their state. It's a bit of an imperfect solution, but since the dataflow changes dynamically as queries arrive, the precise column types aren't known at compile time. One constraint for doing fancy things with `DataType` (like adding signedness as a property) is that the memory size of the `DataType` enum is pretty crucial to Noria's memory footprint: every record cell value in Noria uses _at least_ `sizeof(DataType)` bytes, so adding additional enum variants or levels of enum indirection can come at some cost.

**Comment by alexsnaps** [comment]
@ms705 I've just opened a PR on ms705/nom-sql to only realize it's your repo 🤦‍♂️ 
Anyways, that's the work I had done (plus on a `v0_0_4` branch to avoid bumping noria) to support unsigned types when parsing the SQL. If we can get that in a reasonable fashion for merge, we can then see how to move this PR forward. wdyt?

**Comment by alexsnaps** [comment]
With ms705/nom-sql#40 merged, I'll rebase and update this PR

**Comment by alexsnaps** [comment]
Alright, I think this "only" misses the arithmetic operators on all the new permutation. I'll add these this week sometime 🙈. But I am missing something else? wdyt?

**Comment by ms705** [comment]
Thanks for coming back to this! :+1:

Released the updated `nom-sql` as v0.0.8. I'll take a look at the changes today or tomorrow.

File: noria/src/data.rs [hunk 1]
```diff
@@ -516,13 +516,20 @@ macro_rules! arithmetic_operation (
...
             (&DataType::BigInt(a), &DataType::Int(b)) => (a $op i64::from(b)).into(),
-            // TODO ADD SUPPORT FOR MATHS ON SIGNED (Big)Int
+            (&DataType::UnsignedBigInt(a), &DataType::Int(b)) => (i128::from(a) $op i128::from(b)).into(),
+            (&DataType::UnsignedBigInt(a), &DataType::UnsignedInt(b)) => (a $op u64::from(b)).into(),
+            (&DataType::UnsignedInt(a), &DataType::UnsignedBigInt(b)) => (u64::from(a) $op b).into(),
 
             (first @ &DataType::Int(..), second @ &DataType::Real(..)) |
+            (first @ &DataType::UnsignedInt(..), second @ &DataType::Real(..)) |
             (first @ &DataType::Real(..), second @ &DataType::Int(..)) |
+            (first @ &DataType::Real(..), second @ &DataType::UnsignedInt(..)) |
```
Comment (alexsnaps) [inline noria/src/data.rs]: No (`Unsigned`)`BigInt` here?

File: noria/src/data.rs [hunk 2]
```diff
@@ -243,21 +269,43 @@ impl Hash for DataType {
     }
 }
 
+impl From<i128> for DataType {
+    fn from(s: i128) -> Self {
+        if s < 0 {
+            DataType::BigInt(s as i64)
```
Comment (ms705) [inline noria/src/data.rs]: Will these types of casts always succeed? There are negative values in the `i128` range that won't fit into an `i64`, right?

(Below cast from the positive `i128` to `u64` should be fine in that regard, though, as `u64` covers the entire positive range of `i128`.)

File: noria/src/data.rs (hunk 1 above)
Comment (ms705) [reply noria/src/data.rs]: The max and min values of `f64` (64-bit floating point) fit inside the `i128` range, so comparison with `BigInt` seems fine. Do you have a reason to think we don't want it here?

Comment (alexsnaps) [reply noria/src/data.rs]: That's why I was surprised (Signed)`BigInt` wasn't there. I'll add it.

File: noria/src/data.rs (hunk 2 above)
Comment (alexsnaps) [reply noria/src/data.rs]: I _think_ this is fine… others than a value coming from a stateful operator… But that would have been an issue before too, right?

Comment (alexsnaps) [reply noria/src/data.rs]: Added a check for `s >= i64::MIN` anyways…

Comment (ms705) [reply noria/src/data.rs]: :+1:

**Comment by ms705** [comment]
Thanks, looks good to merge to me once CI has finished.

**Pull request merged**